modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C

```diff
@@ -0,0 +1,627 @@
+    _models[i]->updateSubstepConstitutiveVariableValues();
+
+  // save off the old F^{p} inverse now that have converged on the stress and state variables
+  _inverse_plastic_deformation_grad_old = _inverse_plastic_deformation_grad;
+}
+
+void
+ComputeMultipleCrystalPlasticityStress::postSolveQp(RankTwoTensor & cauchy_stress,
+                                                    RankFourTensor & jacobian_mult)
+{
+  cauchy_stress = _elastic_deformation_gradient * _pk2[_qp] *
+                  _elastic_deformation_gradient.transpose() / _elastic_deformation_gradient.det();
+
+  calcTangentModuli(jacobian_mult);
+
+  _total_lagrangian_strain[_qp] =
+      _deformation_gradient[_qp].transpose() * _deformation_gradient[_qp] -
+      RankTwoTensor::Identity();
+  _total_lagrangian_strain[_qp] = _total_lagrangian_strain[_qp] * 0.5;
+
+  // Calculate crystal rotation to track separately
+  RankTwoTensor rot;
+  _deformation_gradient[_qp].getRUDecompositionRotation(rot);
+  _update_rotation[_qp] = rot * _crysrot[_qp];
+}
+
+void
+ComputeMultipleCrystalPlasticityStress::solveStateVariables()
+{
+  unsigned int iteration;
+  bool iter_flag = true;
+
+  iteration = 0;
+  // Check for slip system resistance update tolerance
+  while (iter_flag && iteration < _maxiterg)
+  {
+    solveStress();
+    if (_error_tolerance)
+      return;
+
+    _plastic_deformation_gradient[_qp] =
+        _inverse_plastic_deformation_grad.inverse(); // the postSoveStress
+
+    // Update slip system resistance and state variable after the stress has been finalized
+    // We loop through all the models for each calculation
+    // in order to make sure that when coupling appears, the state variables are updated based on
+    // the same components
+    for (unsigned int i = 0; i < _num_models; ++i)
+      _models[i]->cacheStateVariablesBeforeUpdate();
+
+    for (unsigned int i = 0; i < _num_models; ++i)
+      _models[i]->calculateStateVariableEvolutionRateComponent();
+
+    for (unsigned int i = 0; i < _num_models; ++i)
+      if (!_models[i]->updateStateVariable())
+        _error_tolerance = true;
+
+    for (unsigned int i = 0; i < _num_models; ++i)
+      _models[i]->calculateSlipResistance();
+
+    if (_error_tolerance)
+      return;
+
+    for (unsigned int i = 0; i < _num_models; ++i)
+    {
+      // returns false if all values are converged and good to go
+      if (_models[i]->areConstitutiveStateVariablesConverged())
+      {
+        iter_flag = true;
+        break;
+      }
+      else
+        iter_flag = false; // iter_flag = false only when all models returns false
+    }
+
+    if (iter_flag)
+    {
+#ifdef DEBUG
+      mooseWarning("ComputeMultipleCrystalPlasticityStress: State variables (or the system "
+                   "resistance) did not "
+                   "converge at element ",
+                   _current_elem->id(),
+                   " and qp ",
+                   _qp,
+                   "\n");
+#endif
+    }
+    iteration++;
+  }
```

Comment:
```suggestion
  unsigned int iteration = 0;
  bool iter_flag;

  // Check for slip system resistance update tolerance
  do
  {
    solveStress();
    if (_error_tolerance)
      return;

    _plastic_deformation_gradient[_qp] =
        _inverse_plastic_deformation_grad.inverse(); // the postSoveStress

    // Update slip system resistance and state variable after the stress has been finalized
    // We loop through all the models for each calculation
    // in order to make sure that when coupling appears, the state variables are updated based on
    // the same components
    for (unsigned int i = 0; i < _num_models; ++i)
      _models[i]->cacheStateVariablesBeforeUpdate();

    for (unsigned int i = 0; i < _num_models; ++i)
      _models[i]->calculateStateVariableEvolutionRateComponent();

    for (unsigned int i = 0; i < _num_models; ++i)
      if (!_models[i]->updateStateVariable())
        _error_tolerance = true;

    for (unsigned int i = 0; i < _num_models; ++i)
      _models[i]->calculateSlipResistance();

    if (_error_tolerance)
      return;

    iter_flag = false;
    for (unsigned int i = 0; i < _num_models; ++i)
    {
      // returns false if all values are converged and good to go
      if (_models[i]->areConstitutiveStateVariablesConverged())
      {
        iter_flag = true;
        break;
      }
    }

    if (iter_flag)
    {
#ifdef DEBUG
      mooseWarning("ComputeMultipleCrystalPlasticityStress: State variables (or the system "
                   "resistance) did not "
                   "converge at element ",
                   _current_elem->id(),
                   " and qp ",
                   _qp,
                   "\n");
#endif
    }
    iteration++;
  } while(iter_flag && iteration < _maxiterg)
```

Reply:
also:

```
      // returns false if all values are converged and good to go
      if (_models[i]->areConstitutiveStateVariablesConverged())
```

This seems backwards to me. Shouldn't it return true with this function name?

Reply:
I agree, it is a quite counterintuitive here. I was puzzled by these error checks before. 

I essentially kept the same convergence checks and error tolerances form the `FiniteStrainUOBasedCP ` for easier comparison and debugging. 

I guess it is time to make the change